File: src/orpheus/extractors/bitcoin_core.py
"""Bitcoin Core wallet.dat extractor.

Works on both BDB and SQLite backends by scanning raw bytes for the DER-encoded
secp256k1 private-key pattern used by Bitcoin Core:

    30 81 D3 02 01 01 04 20 <32 bytes>

Encrypted (password-protected) wallets won't yield raw keys from this scan —
they appear as AES-CBC ciphertext keyed by the wallet password. For those we
rely on the `encrypted` extractor path (which can wrap this one).
"""

from __future__ import annotations

from pathlib import Path

from orpheus.crypto.addresses import all_addresses_for_privkey
from orpheus.crypto.keys import privkey_to_wif, validate_privkey
from orpheus.extractors.base import Extractor, register
from orpheus.models import ExtractedKey, SourceType

DER_PATTERN = b"\x30\x81\xd3\x02\x01\x01\x04\x20"
PRIVKEY_LEN = 32


@register
class BitcoinCoreExtractor(Extractor):
    source_type = SourceType.BITCOIN_CORE

    def can_handle(self, path: Path) -> bool:
        if not path.is_file():
            return False
        name = path.name.lower()
        # Only claim files that look like Bitcoin Core wallets — prefix name checks
        # aren't enough because MultiBit also uses *.wallet.
        if not (name.endswith(".dat") or name == "wallet"):
            return False
        try:
            head = path.read_bytes()[:4096]
        except OSError:
            return False
        return DER_PATTERN in head or b"main\x00" in head

    def extract(self, path: Path, passwords: list[str] | None = None) -> list[ExtractedKey]:
        data = path.read_bytes()
        keys: list[ExtractedKey] = []
        seen: set[bytes] = set()
        pos = 0
        while True:
            idx = data.find(DER_PATTERN, pos)
            if idx == -1:
                break
            start = idx + len(DER_PATTERN)
            privkey = data[start:start + PRIVKEY_LEN]
            pos = idx + 1
            if len(privkey) != PRIVKEY_LEN or privkey in seen or not validate_privkey(privkey):
                continue
            seen.add(privkey)
            keys.append(_privkey_to_extracted(privkey, str(path)))
        return keys


def _privkey_to_extracted(privkey: bytes, source_file: str) -> ExtractedKey:
    addrs = all_addresses_for_privkey(privkey)
    return ExtractedKey(
        wif=privkey_to_wif(privkey, compressed=True),
        address_compressed=addrs["p2pkh_compressed"],
        address_uncompressed=addrs["p2pkh_uncompressed"],
        address_p2sh_segwit=addrs["p2sh_p2wpkh"],
        address_bech32=addrs["bech32"],
        source_file=source_file,
        source_type=SourceType.BITCOIN_CORE,
    )

File: crates/orpheus-core/src/extractors/bitcoin_core.rs
//! Bitcoin Core wallet.dat (BDB or SQLite) — DER-pattern private-key scan.

use std::path::Path;

use crate::{
    crypto::{addresses_for_privkey, privkey_to_wif},
    extractors::{Extractor, scan_result_error},
    models::{ExtractedKey, SourceType, WalletScanResult},
};

pub const DER_PATTERN: &[u8] = b"\x30\x81\xd3\x02\x01\x01\x04\x20";
const PRIVKEY_LEN: usize = 32;

pub struct BitcoinCoreExtractor;

impl Extractor for BitcoinCoreExtractor {
    fn source_type(&self) -> SourceType {
        SourceType::BitcoinCore
    }

    fn can_handle(&self, path: &Path) -> bool {
        if !path.is_file() {
            return false;
        }
        let name = path
            .file_name()
            .and_then(|n| n.to_str())
            .unwrap_or("")
            .to_lowercase();
        if !(name.ends_with(".dat") || name == "wallet") {
            return false;
        }
        let Ok(head) = read_head(path, 4096) else { return false };
        memmem(&head, DER_PATTERN) || memmem(&head, b"main\0")
    }

    fn extract(&self, path: &Path, _passwords: &[String]) -> WalletScanResult {
        let data = match std::fs::read(path) {
            Ok(d) => d,
            Err(e) => return scan_result_error(path, self.source_type(), e),
        };
        let mut keys = Vec::new();
        let mut seen = std::collections::HashSet::new();
        let mut pos = 0usize;
        while pos + DER_PATTERN.len() + PRIVKEY_LEN <= data.len() {
            if &data[pos..pos + DER_PATTERN.len()] == DER_PATTERN {
                let start = pos + DER_PATTERN.len();
                let priv_bytes: [u8; 32] = match data[start..start + PRIVKEY_LEN].try_into() {
                    Ok(b) => b,
                    Err(_) => {
                        pos += 1;
                        continue;
                    }
                };
                if seen.insert(priv_bytes) {
                    if let Some(k) = make_key(&priv_bytes, path, SourceType::BitcoinCore) {
                        keys.push(k);
                    }
                }
            }
            pos += 1;
        }
        WalletScanResult {
            source_file: path.display().to_string(),
            source_type: self.source_type(),
            keys,
            error: None,
        }
    }
}

pub(crate) fn read_head(path: &Path, max: usize) -> std::io::Result<Vec<u8>> {
    use std::io::Read;
    let mut f = std::fs::File::open(path)?;
    let mut buf = vec![0u8; max];
    let n = f.read(&mut buf)?;
    buf.truncate(n);
    Ok(buf)
}

pub(crate) fn memmem(hay: &[u8], needle: &[u8]) -> bool {
    if needle.is_empty() || needle.len() > hay.len() {
        return false;
    }
    hay.windows(needle.len()).any(|w| w == needle)
}

pub(crate) fn make_key(
    privkey: &[u8; 32],
    source_file: &Path,
    source_type: SourceType,
) -> Option<ExtractedKey> {
    let wif = privkey_to_wif(privkey, true).ok()?;
    let a = addresses_for_privkey(privkey).ok()?;
    Some(ExtractedKey {
        wif,
        address_compressed: a.p2pkh_compressed,
        address_uncompressed: Some(a.p2pkh_uncompressed),
        address_p2sh_segwit: Some(a.p2sh_p2wpkh),
        address_bech32: Some(a.bech32),
        source_file: source_file.display().to_string(),
        source_type,
        derivation_path: None,
        balance_sat: None,
        total_received_sat: None,
        tx_count: None,
        notes: None,
    })
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::io::Write;

    fn priv_hex(s: &str) -> [u8; 32] {
        let mut out = [0u8; 32];
        hex::decode_to_slice(s, &mut out).unwrap();
        out
    }

    #[test]
    fn extracts_planted_der_keys() {
        let tmp = tempfile_path("demo_bc.dat");
        let mut f = std::fs::File::create(&tmp).unwrap();
        f.write_all(&[0u8; 128]).unwrap();
        f.write_all(b"main\0").unwrap();
        let p1 = priv_hex("0c28fca386c7a227600b2fe50b7cae11ec86d3bf1fbe471be89827e19d72aa1d");
        let p2 = priv_hex("18e14a7b6a307f426a94f8114701e7c8e774e7f9a47e2c2035db29a206321725");
        for p in [p1, p2] {
            f.write_all(&[0u8; 16]).unwrap();
            f.write_all(DER_PATTERN).unwrap();
            f.write_all(&p).unwrap();
            f.write_all(&[0u8; 32]).unwrap();
        }
        drop(f);

        let ex = BitcoinCoreExtractor;
        assert!(ex.can_handle(&tmp));
        let result = ex.extract(&tmp, &[]);
        assert_eq!(result.keys.len(), 2);
        let wifs: std::collections::HashSet<_> =
            result.keys.iter().map(|k| k.wif.clone()).collect();
        assert!(wifs.contains("KwdMAjGmerYanjeui5SHS7JkmpZvVipYvB2LJGU1ZxJwYvP98617"));
        std::fs::remove_file(&tmp).ok();
    }

    pub(crate) fn tempfile_path(name: &str) -> std::path::PathBuf {
        let mut p = std::env::temp_dir();
        p.push(format!(
            "orpheus-test-{}-{}",
            std::process::id(),
            name
        ));
        p
    }
}
